packages/backend-data/src/convert_js_resolvers.ts

```diff
@@ -0,0 +1,114 @@
+ */
+const resolveEntryPath = (entry: JsResolverEntry): string => {
+  const unresolvedImportLocationError = new Error(
+    'Could not determine import path to construct absolute code path from relative path. Consider using an absolute path instead.'
+  );
+
+  if (typeof entry === 'string') {
+    return entry;
+  }
+
+  const filePath = new FilePathExtractor(entry.importLine).extract();
+  if (filePath) {
+    return join(dirname(filePath), entry.relativePath);
+  }
+
+  throw unresolvedImportLocationError;
+};
+
+/**
+ * Reads default JS resolver template file and returns string contents sans inline sourcemap
+ *
+ * This returns the top-level passthrough resolver request/response handler (see: https://docs.aws.amazon.com/appsync/latest/devguide/resolver-reference-overview-js.html#anatomy-of-a-pipeline-resolver-js)
+ * It's required for defining a pipeline resolver. The only purpose it serves is returning the output of the last function in the pipeline back to the client.
+ *
+ * Customer-provided handlers are added as a Functions list in `pipelineConfig.functions`
+ */
+const normalizedDefaultJsResolver = (): string => {
+  const resolvedTemplatePath = resolve(
+    fileURLToPath(import.meta.url),
+    '../../lib',
+    JS_PIPELINE_RESOLVER_HANDLER
+  );
+  const fileContents: string = readFileSync(resolvedTemplatePath, 'utf-8');
+  const fileLines = fileContents.split(EOL);
+
+  if (fileLines.pop()?.includes('sourceMappingURL')) {
+    return fileLines.join(EOL);
+  }
+
+  return fileContents;
+};
+
+/**
+ * Converts JS Resolver definition emitted by data-schema into AppSync pipeline
+ * resolvers via L1 construct.
+ */
+export const convertJsResolverDefinition = (
+  scope: Construct,
+  amplifyApi: AmplifyData,
+  jsResolvers: JsResolver[] | undefined
+): void => {
+  if (!jsResolvers || jsResolvers.length < 1) {
+    return;
+  }
+
+  for (const resolver of jsResolvers) {
+    const functions: string[] = resolver.handlers.map((handler, idx) => {
+      const fnName = `Fn_${resolver.typeName}_${resolver.fieldName}_${idx + 1}`;
+      const s3AssetName = `${fnName}_asset`;
+
+      const asset = new Asset(scope, s3AssetName, {
+        path: resolveEntryPath(handler.entry),
+      });
+
+      const fn = new CfnFunctionConfiguration(scope, fnName, {
+        apiId: amplifyApi.apiId,
+        dataSourceName: handler.dataSource,
+        name: fnName,
+        codeS3Location: asset.s3ObjectUrl,
+        runtime: {
+          name: APPSYNC_JS_RUNTIME_NAME,
+          runtimeVersion: APPSYNC_JS_RUNTIME_VERSION,
+        },
+      });
+      fn.node.addDependency(amplifyApi);
+      return fn.attrFunctionId;
+    });
+
+    const resolverName = `Resolver_${resolver.typeName}_${resolver.fieldName}`;
+
+    new CfnResolver(scope, resolverName, {
+      apiId: amplifyApi.apiId,
+      fieldName: resolver.fieldName,
+      typeName: resolver.typeName,
+      kind: APPSYNC_PIPELINE_RESOLVER,
+      code: normalizedDefaultJsResolver(),
```

Comment:
maybe we should use an asset for this as well? Even though it's small, if they have a lot of js resolvers, duplicating this over and over in the template would make it unnecessarily large. We can reuse the same asset in each resolver here.

Reply:
Good call. Updated